kafkaesk/retry.py

```diff
@@ -0,0 +1,354 @@
+from typing import Dict
+from typing import List
+from typing import Optional
+from typing import Tuple
+from typing import Type
+from typing import TYPE_CHECKING
+
+import asyncio
+import logging
+
+if TYPE_CHECKING is True:
+    from .app import Application
+    from .app import Subscription
+
+
+class Record(BaseModel):
+    topic: str
+    partition: int
+    offset: int
+    timestamp: int
+    timestamp_type: int
+    key: Optional[str] = None
+    value: str
+    checksum: Optional[int] = None
+    serialized_key_size: int
+    serialized_value_size: int
+    headers: tuple
+
+    @classmethod
+    def from_consumer_record(cls, record: ConsumerRecord) -> "Record":
+        return cls(
+            topic=record.topic,
+            partition=record.partition,
+            offset=record.offset,
+            timestamp=record.timestamp,  # type: ignore
+            timestamp_type=record.timestamp_type,  # type: ignore
+            key=record.key,  # type: ignore
+            value=record.value.decode("utf-8"),  # type: ignore
+            checksum=record.checksum,  # type: ignore
+            serialized_key_size=record.serialized_key_size,  # type: ignore
+            serialized_value_size=record.serialized_value_size,  # type: ignore
+            headers=record.headers,
+        )
+
+    def to_consumer_record(self) -> ConsumerRecord:
+        # We need to convert the value back into bytes before giving this back to the consumer
+        data = self.dict()
+        data["value"] = data["value"].encode("utf-8")
+
+        return ConsumerRecord(**data)  # type: ignore
+
+
+class FailureInfo(BaseModel):
+    exception: str
+    handler_key: str
+    timestamp: datetime
+
+
+class RetryHistory(BaseModel):
+    failures: List[FailureInfo] = []
+
+
+class RetryMessage(BaseModel):
+    original_record: Record
+    retry_history: RetryHistory
+
+
+class RetryPolicy:
+    logger = logging.getLogger("kafkaesk.retry.retry_policy")
+
+    def __init__(
+        self,
+        app: "Application",
+        subscription: "Subscription",
+        retry_handlers: Optional[Dict[Type[Exception], "RetryHandler"]] = None,
+    ):
+        self.app = app
+        self.subscription = subscription
+
+        self._default_handler: "RetryHandler" = Raise()
+
+        self._handlers = retry_handlers or {}
+        self._handler_cache: Dict[Type[Exception], Tuple[str, RetryHandler]] = {}
+
+        self._ready = False
+
+        if "RetryMessage:1" not in self.app.schemas:
+            self.app.schema("RetryMessage", version=1)(RetryMessage)
+
+    async def add_retry_handler(self, exception: Type[Exception], handler: "RetryHandler") -> None:
+        if exception in self._handlers:
+            raise ValueError(f"{exception} retry handler is already set")
+
+        self._handlers[exception] = handler
+        self.logger.debug(
+            f"Retry policy added new handler {handler.__class__.__name__} for {exception}"
+        )
+
+        if self._ready is True:
```

Comment:
think we need asyncio.Lock implementation to protect `_ready` everywhere it is used?

Reply:
I think we are ok...  I'm using `_ready` more like a `threading.Event` object: it does not set to true before all handlers have been initialized and when finalizing we set the boolean to `False` and then await for any handler cleanup code to finish and return.

Am I missing something that would cause a race here?

Reply:
yah, it's probably fine now. It was only a potential problem when you had the api method to add more handlers so running both at same time could be a problem.

Any time stateful instance variables are used with async code that can be called from multiple places, need to consider lock.